### PR TITLE
Normalize core profile dependencies in web profile

### LINKED ISSUE #138: Make Full / Web profile play nice with MicroProfile maven dependencies
Problem: Jakarta EE API dependencies do not play well with MicroProfile dependencies, even after the empty platform JARs are implemented

Let's assume that Jakarta EE 10.0.1 is a new API version of Jakarta EE 10 with #137 merged in (empty platform JARs),
and some other minor API improvements, just to set up a realistic scenario.
```
<dependency>
    <groupId>jakarta.platform</groupId>
    <artifactId>jakarta.jakartaee-api</artifactId>
    <version>10.0.1</version>
    <scope>provided</scope>
</dependency>
<dependency>
    <groupId>org.eclipse.microprofile</groupId>
    <artifactId>microprofile</artifactId>
    <version>6.0</version>
    <type>pom</type>
    <scope>provided</scope>
</dependency>
```

This will yield the following dependency tree:
```
# mvn dependency:tree
[INFO] +- jakarta.platform:jakarta.jakartaee-api:jar:10.0.1:provided
[INFO] |  +- jakarta.platform:jakarta.jakartaee-web-api:jar:10.0.1:provided
[INFO] |  |  +- jakarta.servlet:jakarta.servlet-api:jar:6.0.0:provided
[INFO] |  |  \- jakarta.security.enterprise:jakarta.security.enterprise-api:jar:3.0.0:provided
[INFO] |  +- jakarta.jms:jakarta.jms-api:jar:3.1.0:provided
[INFO] |  +- jakarta.activation:jakarta.activation-api:jar:2.1.0:provided
[INFO] |  +- jakarta.mail:jakarta.mail-api:jar:2.1.0:provided
...
[INFO] +- org.eclipse.microprofile:microprofile:pom:6.0:provided
[INFO] |  +- jakarta.platform:jakarta.jakartaee-core-api:jar:10.0.0:provided
[INFO] |  +- org.eclipse.microprofile.config:microprofile-config-api:jar:3.0.2:provided
...
```
As you can see above, even though the Jakarta EE API version is 10.0.1, the core profile is still 10.0.0, which may lead to subtle errors that are not readily apparent.

Proposed Solution:
As follow-up to #133, I think we can make web profile depend on core profile, thus normalizing APis.
Currently, there is a whole bunch of copy-and-paste dependencies between core profile and web profile.

With the above solution, the Core Profile dependency that's expected by the Maven resolution rules (10.0.1) would be pulled in.
Even better, if a future version of MicroProfile provides a later version of Core Profile (and it's APIs), that would also work as expected, if the MicroProfile dependency would be put first in the pom.xml. This is also consistent with maven resolution rules, and thus would be exactly what developers would expect.

[AB#91](https://dev.azure.com/jakarta-ee-azdo/7cf9a9d3-9854-41e0-9a66-99506a9ed874/_workitems/edit/91)

### PR DESCRIPTION
**Pull request by lprimak**
## Summary
Conclusion of #133 Bug fix train
Makes Core profile a true dependency of Web profile and plays nice with MicroProfile POM
fixes #138 

## Goals
* Make `jakartaee-api` and `jakartaee-web-api` depend on `jakartaee-core-api` to be consistent with maven resolution rules, and thus make it intuitive to application both app devs and MicroProfile.
* Remove `*-api.cp.version` properties, as they are now overridable with 'regular' maven pom directives, and thus no longer necessary. 
* Remove duplicate dependencies from `jakartaee-web-api` that are already in `jakartaee-core-api` (this can now be done safely because empty JARs are now shipped.
* Make the minimum changes necessary to accomplish the goal.
* Make this suitable as a patch release for Jakarta EE 10 (ex: 10.0.1)
* Remove unnecessary `<optional>` tags from POMs
* Preserve generation of javadoc as it is now (jakarta-api includes platrorm, web and core javadocs, jakartaee-web-api includes web and core javadocs, etc)
* Make javadoc generation errors fail the build. Otherwise, they fail silently generating invalid javadocs (possibly leading to invalid releases)

## Non-Goals
* Not intended as a complete refactoring of the POM files

### CODE REVIEW

**Comment by lprimak** [comment]
@Emily-Jiang This is the PR that you have concerns with.
Thank you!

**Comment by Emily-Jiang** [comment]
> @Emily-Jiang This is the PR that you have concerns with. Thank you!

I did a quick troll through the core profile pom.xml and you were right that the CDI dependency did not have exclusion. However, I have concerns with the core profile without any exclusion for CDI as in MicroProfile, we specified exclusion deliberately. See [here](https://github.com/eclipse/microprofile-parent/blob/2.x/pom.xml#L149). @starksm64 I thought the EL and EJB dependencies were excluded from Core Profile. Why was it not the case?

**Comment by lprimak** [comment]
Emily, please see https://github.com/jakartaee/jakartaee-api/blob/10c589401e5644661b786607c1fd2625f306a270/jakartaee-core-api/pom.xml#L111
In core profile, cdi-api excludes all dependencies, so I *think* what you are saying is correct. Both in current core API and after this PR is merged.

**Comment by lprimak** [comment]
Thanks for looking into this Emily!
Also, just FYI the reason I started looking into this is that Jakarta EE APIs as they currently are not playing nice with MicoProfile maven dependencies. After these PRs are merged, both Jakarta EE and MP will play much nicer with each other!

**Comment by starksm64** [comment]
> > @Emily-Jiang This is the PR that you have concerns with. Thank you!
> 
> @starksm64 I thought the EL and EJB dependencies were excluded from Core Profile. Why was it not the case?

The exclusion was at the spec level and could not be done at the api jar level due to the fact that we were rebuilding the jar from source. With the move to using empty jars these can be excluded.

File: jakartaee-api/pom.xml
Comment (edburns) [inline jakartaee-api/pom.xml:97]: @JanWesterkamp-iJUG questions the correctness of having this here. Remove the `<optional>false</optional>`.

**Comment by arjantijms** [comment]
See also https://github.com/jakartaee/cdi/issues/640

File: jakartaee-api/pom.xml
Comment (JanWesterkamp-iJUG) [inline jakartaee-api/pom.xml:92]: If the intention this to win in the resolution mechanism, then it should be defined before the Web Profile dependency.

**Review by JanWesterkamp-iJUG — CHANGES_REQUESTED** [review]

File: jakartaee-bom/pom.xml
Comment (JanWesterkamp-iJUG) [inline jakartaee-bom/pom.xml:1]: I have two findings here:

The BOM misses a separate Core Profile list of dependencies, that should be defined first before the Web Profile ones, so they can be managed easier.

We should get rid of this plugin here, may be in a separate PR:

        <plugins>
            <plugin>
                <groupId>org.glassfish.build</groupId>
                <artifactId>glassfishbuild-maven-plugin</artifactId>
                <executions>
                    <execution>
                        <id>generate-pom</id>
                        <goals>
                            <goal>generate-pom</goal>
                        </goals>
                    </execution>
                    <execution>
                        <id>attach-all-artifacts</id>
                        <goals>
                            <goal>attach-all-artifacts</goal>
                        </goals>
                    </execution>
                </executions>
            </plugin>

**Review by JanWesterkamp-iJUG — CHANGES_REQUESTED** [review]

File: jakartaee-api/pom.xml
Comment (lprimak) [reply jakartaee-api/pom.xml:1]: Probably... yes this is a separate change. I think there is already an existing issue for htat

File: jakartaee-api/pom.xml
Comment (lprimak) [reply jakartaee-api/pom.xml:92]: Yes, the comment is a bit misleading. The intent is to be "neutral" aka "carry through" the user's intent. If the user wants "MP's version of Jakarta Core Profile" to win, they can put the MP dependency first. If they want Jakarta's Core to win, they put Jakarta first.

This is to be consistent and intuitive with how maven resolution works, and able to specify exactly what the user intended

File: jakartaee-bom/pom.xml
Comment (lprimak) [reply jakartaee-bom/pom.xml:1]: yes, great idea. Also probably not part of this particular PR though

Comment (lprimak) [reply jakartaee-bom/pom.xml:1]: @JanWesterkamp-iJUG If you want to create a separate issue for that, I can volunteer to create a PR there too

File: jakartaee-api/pom.xml
Comment (lprimak) [reply jakartaee-api/pom.xml:92]: I will try to find a better language for the comment

File: jakartaee-web-api/pom.xml
Comment (JanWesterkamp-iJUG) [inline jakartaee-web-api/pom.xml:241]: This is normal intended behaviour:

As documented (https://maven.apache.org/guides/introduction/introduction-to-optional-and-excludes-dependencies.html) optional dependencies are not included transitional, so they need to decalred again, when used.

**Review by JanWesterkamp-iJUG — CHANGES_REQUESTED** [review]
May be we can change the comment reflecting this behaviour as explanation.

File: jakartaee-api/pom.xml
Comment (lprimak) [reply jakartaee-api/pom.xml:93]: Ok, I found better language for the comment and committed it. Thank you!

Comment (JanWesterkamp-iJUG) [reply jakartaee-api/pom.xml:93]: Or just move it up one dependency?

1. Core Profile dependency
2. Web Profile dependency
3. all the rest beloging to Platform

**Review by JanWesterkamp-iJUG — CHANGES_REQUESTED** [review]

File: jakartaee-api/pom.xml
Comment (JanWesterkamp-iJUG) [reply jakartaee-api/pom.xml]: Looks like you catched all the `<optional>false</optional>` statements so far.

**Comment by lprimak** [comment]
@starksm64 Would love your feedback here... thank you!

File: pom.xml
Comment (Emily-Jiang) [inline pom.xml:242]: Have you tried to generate javadoc to ensure no javadoc failures and this won't break build?

File: jakartaee-web-api/pom.xml
Comment (Emily-Jiang) [reply jakartaee-web-api/pom.xml:185]: I see. You deleted the previous exclusion block.

**Review by Emily-Jiang — APPROVED** [review]

File: pom.xml
Comment (Emily-Jiang) [reply pom.xml:242]: I finally found out where you addressed this comment via your email explanation. Thanks

**Comment by lprimak** [comment]
Thank you @Emily-Jiang !
Still waiting for feedback from @JanWesterkamp-iJUG and @starksm64

**Review by starksm64 — APPROVED** [review]

**Comment by lprimak** [comment]
Thank you Scott!
@JanWesterkamp-iJUG you are the last one :)

**Comment by lprimak** [comment]
@JanWesterkamp-iJUG ping please :)

File: jakartaee-api/pom.xml
Comment (JanWesterkamp-iJUG) [reply jakartaee-api/pom.xml:93]: In an ideal world, where we could assume that there is only one single version for a spec defined as a depedency in the Profiles and the component specs themself, we would not need to reference the Core Profile here again.

In the situation, that the Web Profile makes ajustments to Core Profile dependencies (like having optional parts becoming required or build time dependencies being not part of the Core Profile set, as mentioned in #140), we need to reference this in the order of having Web Profile picked first, then having Core Profile as a second regarding the versions picked and need to verify that the additional configuration is picked (and kept) from the Web Profile too (and is not overwritten).

The comment sould point to that fact, so we know when we could remove the Core Profile dependency in a future release (where the issues mentioned above fixed) or to prevent somebody to change the order. 
@lprimak do you have a suggestion for the comment? Otherwise I can give a try.

File: jakartaee-web-api/pom.xml
Comment (JanWesterkamp-iJUG) [inline jakartaee-web-api/pom.xml:254]: I think, this should become a required part in the Core (or Web Profile) in a future release (Jakarta EE 11) because XML is still in heavy use and is also required for some configurations (like Persistence and optionally in Servlet) - but that's a separate issue.

Good to fix it here now! But is the exclusion required - I don't think so. But I can verify this with a jQA run, to be sure.

**Review by JanWesterkamp-iJUG — CHANGES_REQUESTED** [review]
@lprimak I added some comments - sorry for the delay.

File: jakartaee-web-api/pom.xml
Comment (JanWesterkamp-iJUG) [reply jakartaee-web-api/pom.xml:254]: Jakarta XML Binding got dependency to Jakarta Activation 2.1.0 - this need to be maintained. So it looks like we need this exclusion until we have that solved first unfortunately.

**Comment by lprimak** [comment]
@JanWesterkamp-iJUG I granted you write permissions for this PR so you can try some language or you can comment here

**Comment by lprimak** [comment]
Also, this PR has been open for 3 months with minimal changes and no issues. Let's merge this in and then we can work on the other, very valid, issues.

Thank you @edburns and @JanWesterkamp-iJUG !

**Comment by edburns** [comment]
Hello @lprimak , I have merged #140 from @lukasj . Can you please revise this PR to take this fact into account?

**Comment by lprimak** [comment]
@edburns done thank you

**Review by JanWesterkamp-iJUG — APPROVED** [review]
As dicussed in the Jakarta Platform call on 2023-08-22 the remaining chages will be handled in separate PRs - so this on could be merged.

Meanwhile #140 was merged before, but this one is rebased and remaining issues in #140 can be done after merging this on first.

@edburns when you agree on this, I will do this merge then.

**Review by edburns — APPROVED** [review]

**Comment by JanWesterkamp-iJUG** [comment]
@edburns thanks for your approval!
I will merge this now.
Further (required) updates will be done is a separate PR, as updating comments or potential updates to optional dependency handling (some are updated in #140).

**Pull request merged**
